### PR TITLE
Avoid causing infinite loop when message is empty

### PR DESCRIPTION
**Pull request by IshitaTakeshi**
Fix https://github.com/ros2/rclpy/issues/934

### CODE REVIEW

**Review by fujitatomoya — COMMENTED** [review]
Could you also address the following? (the same problem)

https://github.com/ros2/rclpy/blob/2db2d3578c7ca7c7742b08efb63e6f0d0ae355e0/rclpy/test/test_client.py#L106-L115

File: rclpy/test/test_node.py [hunk 1]
```diff
@@ -157,8 +157,7 @@ def test_take(self):
             1)
         basic_types_msg = BasicTypes()
         basic_types_pub.publish(basic_types_msg)
-        cycle_count = 0
-        while cycle_count < 5:
+        for i in range(5):
```
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply rclpy/test/test_node.py:160]: yes, i thought that would be flaky for some platform, maybe we could have more buffer here.

[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply rclpy/test/test_node.py:160]: Resolved in c14ef93e587a81de3eec2c7a0e39867d2128cf80

[the remark this answers is not in the document]
Comment (IshitaTakeshi) [reply rclpy/test/test_node.py:160]: Resolved in b967d1c9957b7ab69f95e604bcf0e0fc31ce5cf3

**Comment by fujitatomoya** [comment]
@IshitaTakeshi could you also fix https://github.com/ros2/rclpy/pull/935#pullrequestreview-959217424? this causes the same infinite loop.

**Review by fujitatomoya — APPROVED** [review]

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16642/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11224/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17040/)

**Review by jacobperron — APPROVED** [review]

**Comment by jacobperron** [comment]
Thanks for the fix!

**Comment by sloretz** [comment]
@mergifyio backport humble

**Pull request merged**